druntime/src/core/sys/linux/linux/if_packet.d

```diff
@@ -0,0 +1,318 @@
+    PACKET_FANOUT_FLAG_DEFRAG          = 0x8000,
+}
+
+struct tpacket_stats
+{
+    uint tp_packets;
+    uint tp_drops;
+}
+
+struct tpacket_stats_v3
+{
+    uint tp_packets;
+    uint tp_drops;
+    uint tp_freeze_q_cnt;
+}
+
+struct tpacket_rollover_stats
+{
+    align(8):
+    ulong tp_all;
+    ulong tp_huge;
+    ulong tp_failed;
+}
+
+union tpacket_stats_u
+{
+    tpacket_stats    stats1;
+    tpacket_stats_v3 stats3;
+}
+
+struct tpacket_auxdata
+{
+    uint   tp_status;
+    uint   tp_len;
+    uint   tp_snaplen;
+    ushort tp_mac;
+    ushort tp_net;
+    ushort tp_vlan_tci;
+    ushort tp_vlan_tpid;
+}
+
+enum : uint
+{
+    TP_STATUS_KERNEL          = 0,
+    TP_STATUS_USER            = 1 << 0,
+    TP_STATUS_COPY            = 1 << 1,
+    TP_STATUS_LOSING          = 1 << 2,
+    TP_STATUS_CSUMNOTREADY    = 1 << 3,
+    TP_STATUS_VLAN_VALID      = 1 << 4,
+    TP_STATUS_BLK_TMO         = 1 << 5,
+    TP_STATUS_VLAN_TPID_VALID = 1 << 6,
+    TP_STATUS_CSUM_VALID      = 1 << 7,
+    TP_STATUS_GSO_TCP         = 1 << 8,
+}
+
+enum : uint
+{
+    TP_STATUS_AVAILABLE    = 0,
+    TP_STATUS_SEND_REQUEST = 1 << 0,
+    TP_STATUS_SENDING      = 1 << 1,
+    TP_STATUS_WRONG_FORMAT = 1 << 2,
+}
+
+enum : uint
+{
+    TP_STATUS_TS_SOFTWARE     = 1 << 29,
+    TP_STATUS_TS_RAW_HARDWARE = 1U << 31,
+}
+
+enum uint TP_FT_REQ_FILL_RXHASH = 0x1;
+
+struct tpacket_hdr
+{
+    ulong   tp_status;
+    uint    tp_len;
+    uint    tp_snaplen;
+    ushort  tp_mac;
+    ushort  tp_net;
+    uint    tp_sec;
+    uint    tp_usec;
+}
+
+enum TPACKET_ALIGNMENT = 16;
+size_t TPACKET_ALIGN(size_t x) { return (x + TPACKET_ALIGNMENT - 1) &~ (TPACKET_ALIGNMENT - 1); }
+enum TPACKET_HDRLEN = TPACKET_ALIGN(tpacket_hdr.sizeof) + sockaddr_ll.sizeof;
+
+struct tpacket2_hdr
+{
+    uint     tp_status;
+    uint     tp_len;
+    uint     tp_snaplen;
+    ushort   tp_mac;
+    ushort   tp_net;
+    uint     tp_sec;
+    uint     tp_nsec;
+    ushort   tp_vlan_tci;
+    ushort   tp_vlan_tpid;
+    ubyte[4] tp_padding;
+}
+
+struct tpacket_hdr_variant1
+{
+    uint   tp_rxhash;
+    uint   tp_vlan_tci;
+    ushort tp_vlan_tpid;
+    ushort tp_padding;
+}
+
+struct tpacket3_hdr
+{
+    uint   tp_next_offset;
+    uint   tp_sec;
+    uint   tp_nsec;
+    uint   tp_snaplen;
+    uint   tp_len;
+    uint   tp_status;
+    ushort tp_mac;
+    ushort tp_net;
+
+    union
+    {
+        tpacket_hdr_variant1 hv1;
+    }
+
+    ubyte[8] tp_padding;
+}
+
+struct tpacket_bd_ts
+{
+    uint ts_sec;
+
+    union
+    {
+        uint ts_usec;
+        uint ts_nsec;
+    }
+}
+
+struct tpacket_hdr_v1
+{
+    uint block_status;
+    uint num_pkts;
+    uint offset_to_first_pkt;
+    uint blk_len;
+    align(8) ulong seq_num;
+    tpacket_bd_ts ts_first_pkt;
+    tpacket_bd_ts ts_last_pkt;
+}
+
+union tpacket_bd_header_u
+{
+    tpacket_hdr_v1 bh1;
+}
+
+struct tpacket_block_desc
+{
+    uint version_;
+    uint offset_to_priv;
+
+    union
+    {
+        tpacket_bd_header_u hdr;
```

Comment:
This shouldn't be wrapped in a union. The tpacket_bd_header_u is a union type.

Reply:
Ah, I got confused by the `union` in front. It's the same thing as with structs in C except that it's a `union`. That makes sense. I haven't been dealing with C much lately and completely misunderstood that.

Reply:
I had thought that it was odd to have a one member union. LOL.

Reply:
fixed